### ISSUE TITLE
@prefresh/vite HMR working erratically

### ISSUE THREAD

**Issue by PuruVJ** [JoviDeCroock/prefresh#294]
Vite version: 2.1.0
@prefresh/vite version: 2.1.0

Ok, so I'm having a very strange issue here. My project(50+ TS/X files)'s HMR seemed to work well before I did something like this:

```tsx
import { PlaceholderApp } from '__/components/apps/Placeholder/Placeholder';
import { createAppConfig } from '__/helpers/create-app-config';

export const viewSourceAppConfig = createAppConfig({
  title: `View Source`,
  resizable: true,
  Component: () => <PlaceholderApp />,

  shouldOpenWindow: false,
  externalAction: () => window.open('https://github.com/puruvj/macos-web', '_blank'),
});
```

Notice Component property here
Now, this config is loaded in another object, and from there, its pulled  into a component, like this:

```tsx
export const Window = ({ appID }: WindowProps) => {
  const { Component, resizable } = appsConfig[appID];

  return (<Component />);
}
```

After I added this, the HMR is just gone. On most changes, the whole page reloads
This is on editing anything, not just this PlaceholderApp or Window component, anything!
And this is quite variable. The change seems to happen randomly. Sometimes, there will be HMR, and sometimes it reloads. It seems to alternate between those two

I think it might be because Vite can't statically analyze the components, but it affecting HMR for whole app in this erratic fashion is just painful.

One of the config files: https://github1s.com/puruvj/macos-web/blob/HEAD/src/data/apps/view-source.app-config.tsx
The object where all these configs go: https://github1s.com/puruvj/macos-web/blob/HEAD/src/data/apps/apps-config.ts
The component where this `Component` is extracted and used: https://github1s.com/puruvj/macos-web/blob/HEAD/src/components/Desktop/Window/Window.tsx

**Comment by JoviDeCroock** [comment]
I wonder if this is an issue with how this update is being bubbled, as this might not be a self-accepting file it will bubble up to a non-boundary.

I can check if this is an issue with Prefresh but I'm not entirely sure if this is a good pattern to HMR from, will report back when I get the time to look into it.

**Comment by PuruVJ** [comment]
> I wonder if this is an issue with how this update is being bubbled, as this might not be a self-accepting file it will bubble up to a non-boundary.
> 
> I can check if this is an issue with Prefresh but I'm not entirely sure if this is a good pattern to HMR from, will report back when I get the time to look into it.

I guessed that this passing components around in configs isn't a great idea either, but I couldn't think up anyway to provide components for each config in any other way either

**Comment by JoviDeCroock** [comment]
As expected, the two `view-source.app-config.tsx` and `apps-config.ts` can't register any components reliably https://astexplorer.net/#/gist/d62c4ecf5d42cd7be6b5a8aa334c8fdf/79ebe1f197daf75fd52e9e72432f12fe4cb64adf which means that these will bubble up to `Window.tsx` where it tries to do a hot-reload. I actually think this should work correctly from a Prefresh perspective.

I'd try logging out [here](https://github.com/JoviDeCroock/prefresh/blob/main/packages/vite/src/index.js#L100) if the error doesn't come up there then something in vite internally is conflicting. If it does it might be something in Prefresh that assumes presence of a certain component-type which isn't present in the tree.

**Comment by PuruVJ** [comment]
Alright, I'll try to log it here

**Comment by PuruVJ** [comment]
The log isn't happening, I put it in both `dist/index.js` as well as `dist/index.mjs`.

Also, another thing I noticed, when I edited the `view-source.app-config.ts`, the logs in console showed this:

![image](https://user-images.githubusercontent.com/47742487/111309397-d9eca580-8681-11eb-97b4-3c3ca5646aeb.png)

These are the 3 place where `appsConfig` is being used, whether the component itself isn't used or not. Apprarently, the change is bubbling to these 3 files, as you said. And these logs are there when the HMR is working. Sometimes, page reloads and there are no logs to see

**Comment by PuruVJ** [comment]
So I gather its a Vite problem, then?

**Comment by JoviDeCroock** [comment]
I _think_ so, I could be wrong but the only way we could be causing a `reload` is if we hit that catch block.

**Comment by PuruVJ** [comment]
Maybe, though these didn't work 
![image](https://user-images.githubusercontent.com/47742487/111310588-45834280-8683-11eb-8f68-bd8185678e8c.png)

Just to be sure, I'll put a log in the try block

**Comment by PuruVJ** [comment]
Alright, this is weird. The logs inside the try aren't being executed either 🤔

**Comment by PuruVJ** [comment]
![image](https://user-images.githubusercontent.com/47742487/111310921-9c891780-8683-11eb-86e0-e0288c5869b6.png)

**Comment by JoviDeCroock** [comment]
That would make sense _if_ vite doesn't consider this hot-updateable. Which would happen if the bubbling up from your updated-file never reaches a Preact component

**Comment by PuruVJ** [comment]
Got it. Do you have any temporary solution in mind for this? Maybe I need to return the right component using if else rather than getting it from the Object?

**Comment by PuruVJ** [comment]
I'm the world's greatest idiot!! The above code wasn't the problem, it never was, the problem was in this `WindowArea.tsx`:
![image](https://user-images.githubusercontent.com/47742487/111353872-b9394580-86ab-11eb-9a89-0babfc6b17fa.png)

`openApps` is a key value pair with all values set as `false` initially, so vite resolves it to be false entirely, and `Window` is never statically analyzed, so its out of the HMR loop.

Now I gotta figure out how to make sure Vite statically analyzes it 🤦‍♀️🤦‍♀️🤦‍♀️

**Comment by JoviDeCroock** [comment]
Glad you found out, is it safe to close this issue? :)

**Comment by PuruVJ** [comment]
> Glad you found out, is it safe to close this issue? :)

Wait, even that didn't fix the problem 😥😥. I'll look into it a bit more to confirm why it is happening. Please keep it open for a while

**Comment by PuruVJ** [comment]
Ok, I have now figured out the problem, and It's with the `PlaceholderApp` component. Doing changes in it is what's causing the reloads. `Window`, on any change is not affecting the HMR. `PlaceholderApp` is. It's very strange.

![image](https://user-images.githubusercontent.com/47742487/111357586-71b4b880-86af-11eb-9713-300edb9e1cb0.png)

Its safe to close the issue, as it is out of the scope of prefresh, I believe

**Comment by PuruVJ** [comment]
Alright, I've tracked the issue down completely, and its impossibly weird. I don't understand why it is happening.

Any external component inside `Window.tsx` is causing the HMR to break. I switched `PlaceholderApp` with other apps too, but same result. Modifying them cause alternate page reloads. This is the relevant `Window.tsx`: https://github1s.com/PuruVJ/macos-web/blob/feat/fix-hmr-remove-component-prop-from-app-configs/src/components/Desktop/Window/Window.tsx

And this is where the `Window` is called: https://github1s.com/PuruVJ/macos-web/blob/feat/fix-hmr-remove-component-prop-from-app-configs/src/components/Desktop/WindowsArea/WindowsArea.tsx

Got any idea what might be going wrong?
